### PR TITLE
Process documentation in type definitions

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Another part of #3611 ready for integration into `develop` branch.

### Important Notes

Test `org.enso.compiler.EnsoCompilerTest.testTestGroup` is ignored as it has problems with source offsets - identifiers don't have the appropriate names due to `Tree.codeRepr()` being _off_. 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md) style guides.
- [x] All code conforms to the `cargo fmt` style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 1]
```diff
@@ -126,16 +127,18 @@ IR.Module translateModule(Tree module) {
...
               bindings = cons(n, bindings);
               var t = translateModuleSymbol(anno.getExpression());
               bindings = cons(t, bindings);
             }
+            case Tree.Documented doc -> {
+              var c = translateComment(doc);
+              bindings = cons(c, bindings);
+              var t = translateModuleSymbol(doc.getExpression());
+              bindings = cons(t, bindings);
+            }
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: What if something is both annotated and documented? I'd use recursion to handle node types that act as wrappers in any order.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 2]
```diff
@@ -356,28 +359,34 @@ private List<IR.DefinitionArgument> translateArgumentsDefinition(java.util.List<
     * @param body the body to be translated
     * @return the [[IR]] representation of `body`
     */
-  private scala.collection.immutable.List<IR> translateTypeBody(java.util.List<Line> block, boolean found) {
-    var ir = block.stream().map(line -> translateTypeBodyExpression(line)).filter(line -> line != null).iterator();
-    return CollectionConverters.asScala(ir).toList();
+  private List<IR> translateTypeBody(java.util.List<Line> block, boolean found) {
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: What's `found` for?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 3]
```diff
@@ -1130,9 +1153,22 @@ IR.Expression translateDecimalLiteral(
...
+    for (var t : txt.getElements()) {
+      var sectionTxt = switch (t) {
+          case TextElement.Section s -> s.getText().codeRepr();
+          case TextElement.Escape e -> switch (e.getToken().codeRepr()) {
+            case "\\n" -> "\n";
+            case "\\t" -> "\t";
+            case "\\r" -> "\r";
+            case "\\f" -> "\f";
+            default -> throw new UnhandledEntity(e, "translateLiteral");
+          };
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: `e.getValue()` returns the value, which is a Unicode codepoint (or -1 if it was invalid). You can append it to the `StringBuilder` with `sb.appendCodePoint()`

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 4]
```diff
@@ -1879,26 +1915,28 @@ def translateInvalid(invalid: AST.Invalid): Expression = {
...
-            sep = "\n";
+  IR.Comment translateComment(Tree.Documented doc) {
+      var msg = new StringBuilder();
+      for (var t : doc.getDocumentation().getElements()) {
+        switch (t) {
+          case TextElement.Section s -> {
+            var txt = s.getText().codeRepr();
+            if (txt.startsWith("#")) {
+              txt = txt.substring(1);
+            }
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: This shouldn't be necessary after #3802.

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java [hunk 5]
```diff
@@ -407,6 +402,31 @@ public void testMultiParameterFunction() throws Exception {
     """);
   }
 
+  @Test
+  @Ignore // shifted positions confused `Tree.codeRepr()` for identifiers
```
Comment (kazcw) [inline engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java]: Fixed in #3808.

**Review by kazcw — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 6]
```diff
@@ -1879,26 +1915,28 @@ def translateInvalid(invalid: AST.Invalid): Expression = {
...
+            if (msg.length() > 0) {
+              msg.append(" ");
+            }
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Whitespace handling in doc comments is currently the cause of IR-equivalence test failures in most source files. The problem is that in the new parser, I implement the [left-trim algorithm defined in the spec](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/04.%20Expressions.md#inline-and-block-text-literals). The old parser didn't trim anything--the indentation that makes the block was included in the string. So, for compatibility we need to add the whitespace back in here, until the old parser is fully replaced and we can switch to the new semantics. Right now `Token` only provides whitespace info as source offsets; I'll add an accessor that makes this easier.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 7]
```diff
@@ -1879,26 +1915,28 @@ def translateInvalid(invalid: AST.Invalid): Expression = {
...
-    return switch (tree) {
-      case Tree.ArgumentBlockApplication comment -> {
-        var doc = new StringBuilder();
-        var sep = "";
-        for (var l : comment.getArguments()) {
-          if (l.getExpression() != null) {
-            doc.append(sep);
-            doc.append(l.getExpression().codeRepr());
-            sep = "\n";
+  IR.Comment translateComment(Tree.Documented doc) {
```
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: If this takes a `DocComment` argument, then you can use it for `TypeConstructorDef` and `Case` documentation fields.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 6 above)
Comment (kazcw) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: I pushed a small commit integrating the new `getWhitespace`. For some reason trimming one space off every line seems to match the old parser's semantics.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Thank you. Getting the indentation right with the old parser is tricky. Every help counts!

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Used in e7b9b501a2ff6640086a86a2d09f8f8ce0d86e53 - thanks for the tip.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Made more recursive in [0086305](https://github.com/enso-org/enso/pull/3806/commits/008630563777d284f8eeedae3fa7b7c7ee7751cb)

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Debris from the past. Removed in [d6fc9e2](https://github.com/enso-org/enso/pull/3806/commits/d6fc9e2c8a3d2afd3b4d66357a9794baa1ff945a).

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Attempted in [da5ad8f](https://github.com/enso-org/enso/pull/3806/commits/da5ad8f479feefdda0d6e46d877a25f9c40e2263), but not used yet. I guess I will need this in the future.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 8]
```diff
@@ -262,28 +230,49 @@ case List(expr) => Some(expr)
...
         );
+        yield cons(t, appendTo);
       }
       case Tree.Function fn -> {
         var methodRef = translateMethodReference(fn.getName(), false);
         var args = translateArgumentsDefinition(fn.getArgs());
         var body = translateExpression(fn.getBody(), false);
 
-        yield new IR$Module$Scope$Definition$Method$Binding(
+        var b = new IR$Module$Scope$Definition$Method$Binding(
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: it's generally easier to understand vars if they are named with something more than a single letter :)

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 9]
```diff
@@ -262,28 +230,49 @@ case List(expr) => Some(expr)
...
         );
+        yield cons(b, appendTo);
       }
-      /*
+      case Tree.Annotated anno -> {
+        var n = new IR$Name$Annotation("@" + anno.getAnnotation().codeRepr(), getIdentifiedLocation(anno), meta(), diag());
+        yield translateModuleSymbol(anno.getExpression(), cons(n, appendTo));
+      }
+      case Tree.Documented doc -> {
+        var c = translateComment(doc, doc.getDocumentation());
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: `c` as in a comment?

**Review by hubertp — APPROVED** [review]
Minor nits regarding variable names

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 9 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: `c` as a comment. `t` as a tree. `b` as a block or binding, `n` as annotation.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 8 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: OK. All single letter variables renamed in [207583e](https://github.com/enso-org/enso/pull/3806/commits/207583e11df6f84e816af2feed0d6d384ef1c42b)

**Pull request merged**